sklearn/metrics/scorer.py

```diff
@@ -44,7 +45,69 @@
 from ..base import is_regressor
 
 
-class _BaseScorer(metaclass=ABCMeta):
+def _method_caller(cache, estimator, method, *args, **kwargs):
+    """Call estimator with method and args and kwargs."""
+    if cache is None:
+        return getattr(estimator, method)(*args, **kwargs)
+
+    try:
+        return cache[method]
+    except KeyError:
+        result = getattr(estimator, method)(*args, **kwargs)
+        cache[method] = result
+        return result
+
+
+class _MultimetricScorer(dict):
+    """Callable dictionary for multimetric scoring used to avoid repeated calls
+    to `predict_proba`, `predict`, and `decision_function`.
+
+    `_MultimetricScorer` will return a dictionary of scores corresponding to
+    the scorers in the dictionary. Note `_MultimetricScorer` can be created
+    with a dictionary with one key.
+    """
+    def __call__(self, estimator, *args, **kwargs):
+        """Evaluate predicted target values."""
+        scores = {}
+        cache = {} if self._use_cache(estimator) else None
+        method_caller = partial(_method_caller, cache)
+
+        for name, scorer in self.items():
+            if isinstance(scorer, _BaseScorer):
+                score = scorer._score(method_caller, estimator,
+                                      *args, **kwargs)
+            else:
+                score = scorer(estimator, *args, **kwargs)
+            scores[name] = score
+        return scores
+
+    def _use_cache(self, estimator):
+        """Return True if using a cache it is beneficial.
+
+        Caching is beneficial when of these conditions hold:
+          - `predict_proba` will be called twice.
+          - `predict` will be called twice.
+          - `decision_function` will be called twice.
+          - `decision_function` and `predict` is called.
+          - `decision_function` and `predict_proba` is called.
+
+        """
+        if len(self) == 1:
+            return False
+
+        counter = Counter([type(v) for v in self.values()])
+
+        if any(counter[known_type] > 1 for known_type in
+               [_PredictScorer, _ProbaScorer, _ThresholdScorer]):
+            return True
+
+        if counter[_ThresholdScorer] and (counter[_PredictScorer] or
```

Comment:
Can you maybe add a comment here to explain that? Are we using ``decision_function`` for ``_PredictScorer`` if ``_ThresholdScorer`` is used?
And we're using ``predict_proba`` for ``_ThresholdScorer`` if ``_ProbaScorer`` is used?